helix-view/src/editor.rs

```diff
@@ -383,18 +383,49 @@ impl Editor {
             .selections
             .remove(&id);
 
-        if close_buffer {
-            // get around borrowck issues
-            let doc = &self.documents[&view.doc];
+        self.tree.remove(id);
+        self._refresh();
+    }
 
-            if let Some(language_server) = doc.language_server() {
-                tokio::spawn(language_server.text_document_did_close(doc.identifier()));
-            }
-            self.documents.remove(&view.doc);
+    pub fn close_document(&mut self, id: DocumentId, force: bool) -> anyhow::Result<()> {
+        let view = view!(self);
+        let doc = &self.documents[&id];
+
+        if !force && doc.is_modified() {
+            anyhow::bail!(
+                "buffer {:?} is modified",
+                doc.relative_path()
+                    .map(|path| path.to_string_lossy().to_string())
+                    .unwrap_or_else(|| "[scratch]".into())
+            );
         }
 
-        self.tree.remove(id);
+        if let Some(language_server) = doc.language_server() {
+            tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        }
+
+        if let Some(alt) = view.last_accessed_doc {
+            self.switch(alt, Action::Replace);
+            let view_mut = view_mut!(self);
+            view_mut.last_accessed_doc = None;
+        } else if self.documents.len() > 1 {
+            let first_other_doc = self
+                .documents
+                .iter()
+                .find_map(|(&other_id, _)| if other_id == id { None } else { Some(id) })
+                .unwrap();
+            self.switch(first_other_doc, Action::Replace);
```

Comment:
Just realized, you also need to handle the case where the same document is open in multiple splits at once. You probably want to just loop over all the views, fix up their `last_accessed_doc` & `jumps` (since any view might have a jump to this doc) then switch them to a different buffer.